# DATA LOADING AND EDA ( INSPECTION AND CURATION )

## DATA LOADING

Storing the HuggingFace 'read' key in this Google Colab notebook's Secrets (the 'key icon' on the left corner of Google Colab) as 'hf_token'

In [ ]:
from google.colab import userdata
hf_token = userdata.get("hf_token")

Utilize stored HuggingFace 'read' key

In [ ]:
from huggingface_hub import login

login(token=hf_token)

Mount Drive and create folder for project to store dataset in cache

In [ ]:
# Mounting Drive

from google.colab import drive

drive.mount("/content/drive")

Mounted at /content/drive


In [ ]:
# Creating 'cache' folder for storage of dataset

import os

project_dir = "/content/drive/MyDrive/agriculture_rag"
os.makedirs(project_dir, exist_ok=True)

print(project_dir)

/content/drive/MyDrive/agriculture_rag


Configure Hugging Face's cache to use that folder

In [ ]:
# Configuring

os.environ["HF_HOME"] = "/content/drive/MyDrive/agriculture_rag/huggingface"

In [ ]:
# Confirming

print(os.environ.get("HF_HOME"))

/content/drive/MyDrive/agriculture_rag/huggingface


***Dataset One(1) : The Base Dataset***

In [ ]:
import pandas as pd

base_df = pd.read_csv('/content/drive/MyDrive/agriculture_rag/base_documents.csv')

In [ ]:
base_df.head()

,document_id,title,topic,crop,agro_zone,text,origin,source_url,license
0,doc_dis_001,Maize nitrogen deficiency symptoms,crop_diseases,maize,sub_humid,Nitrogen deficiency in maize shows as uniform ...,synthetic,NaN,CC0-1.0
1,doc_dis_002,Common bean rust management,crop_diseases,beans,highland,Bean rust appears as small reddish-brown pustu...,synthetic,NaN,CC0-1.0
2,doc_dis_003,Cassava mosaic disease,crop_diseases,cassava,sub_humid,"Cassava mosaic causes leaf distortion, chlorot...",synthetic,NaN,CC0-1.0
3,doc_pes_001,Fall armyworm in maize,pests,maize,semi_arid,Fall armyworm larvae feed in the whorl leaving...,synthetic,NaN,CC0-1.0
4,doc_pes_002,Stem borer in sorghum,pests,sorghum,semi_arid,Stem borers tunnel inside sorghum stalks causi...,synthetic,NaN,CC0-1.0


***Dataset Two(2) : The CGIAR Source***

In [ ]:
# Load the CGIAR dataset from hugging face
# already download onto Drive

# Remove the triple quotations when ready for initial execution or
# final submission and delete this particular comment after final submission

'''

from datasets import load_dataset

cgiar_docs = load_dataset("CGIAR/gardian-cigi-ai-documents")
print(cgiar_docs)
print(cgiar_docs['train'][0])

'''

'\n\nfrom datasets import load_dataset\n\ncgiar_docs = load_dataset("CGIAR/gardian-cigi-ai-documents")\nprint(cgiar_docs)\nprint(cgiar_docs[\'train\'][0])\n\n'

In [ ]:
# Saving to Google Drive.

# Remove the triple quotations below when ready for initial execution or
# final submission and delete this particular comment after final submission

'''

cgiar_docs.save_to_disk( "/content/drive/MyDrive/agriculture_rag/cgiar_dataset")

'''

'\n\ncgiar_docs.save_to_disk( "/content/drive/MyDrive/agriculture_rag/cgiar_dataset")\n\n'

In [ ]:
# Checking existence of file in Google Drive

!ls -lah /content/drive/MyDrive/agriculture_rag/cigar_dataset

ls: cannot access '/content/drive/MyDrive/agriculture_rag/cigar_dataset': No such file or directory


In [ ]:
os.getcwd()

'/content'

## EXPLORATION DATA ANALYSIS : INSPECTION AND CURATION

### **Dataset Inspection**

In [ ]:
# load dataset from drive

from datasets import load_from_disk

cgiar_docs = load_from_disk(
    "/content/drive/MyDrive/agriculture_rag/cgiar_dataset"
)

print(cgiar_docs)

DatasetDict({
    train: Dataset({
        features: ['metadata', 'keywords', 'sieverID', 'pagecount', 'content', 'tokenCount', 'images', 'tables'],
        num_rows: 68745
    })
})


In [ ]:
train_docs = cgiar_docs['train']
print (train_docs)

Dataset({
    features: ['metadata', 'keywords', 'sieverID', 'pagecount', 'content', 'tokenCount', 'images', 'tables'],
    num_rows: 68745
})


In [ ]:
# Peeping at the first row of data

print (train_docs[0])

{'metadata': {'gardian_id': 'a5bf57d530bc56223e9e17b3f8aa149d', 'source': 'gardian_index', 'url': 'https://cgspace.cgiar.org/rest/bitstreams/9aa6310a-7b0a-419b-8099-e3807b25c0ad/retrieve', 'description': 'Climate information is now an agricultural input just like seeds, fertilizers or equipment which are at the basis of production. In a nutshell, this is what emerges from this assessment report on the CGIAR Program on Climate Change, Agriculture and Food Security (CCAFS) on the transmission of climate information (CI) and relevant agricultural advice to farmers in Kaffrine, Diourbel, Thies, Louga and Fatick administrative regions. Through this project, CCAFS seeks to contribute to improving the resilience of rural people to climate change by mainstreaming climate information more adequately in the planning and implementation of development activities.\nThanks to the involvement of the network of community rural radio stations (URAC), CCAFS estimates that two million people have had acc

Note : The metadata is a feature that contains other useful features. We can see it as what I would call a ***parent-feature*** (feature of features).

In [ ]:
print(train_docs[0]["metadata"])

{'gardian_id': 'a5bf57d530bc56223e9e17b3f8aa149d', 'source': 'gardian_index', 'url': 'https://cgspace.cgiar.org/rest/bitstreams/9aa6310a-7b0a-419b-8099-e3807b25c0ad/retrieve', 'description': 'Climate information is now an agricultural input just like seeds, fertilizers or equipment which are at the basis of production. In a nutshell, this is what emerges from this assessment report on the CGIAR Program on Climate Change, Agriculture and Food Security (CCAFS) on the transmission of climate information (CI) and relevant agricultural advice to farmers in Kaffrine, Diourbel, Thies, Louga and Fatick administrative regions. Through this project, CCAFS seeks to contribute to improving the resilience of rural people to climate change by mainstreaming climate information more adequately in the planning and implementation of development activities.\nThanks to the involvement of the network of community rural radio stations (URAC), CCAFS estimates that two million people have had access to climat

Knowing the datatypes of all eight(8) fields :

In [ ]:
print(train_docs.features)

{'metadata': {'gardian_id': Value('string'), 'source': Value('string'), 'url': Value('string'), 'description': Value('string'), 'id': Value('string'), 'title': Value('string'), 'language': Value('string'), 'resource_type': Value('string'), 'release_year': Value('string'), 'rights': Value('string'), 'dataNODE_id': Value('string'), 'keywords': List(Value('string')), 'geography': {'regions': List(Value('string')), 'countries': List(Value('string'))}}, 'keywords': Value('string'), 'sieverID': Value('string'), 'pagecount': Value('string'), 'content': Value('string'), 'tokenCount': Value('string'), 'images': List(Value('string')), 'tables': List(Value('string'))}


To form the features useful for building our RAG system and question-answering SLM, 'metadata' and 'content' seem to be the interesting and important fields to pay attention to.

So, let's take a deeper look into them.

Inspecting the keys and values of the features and particularly 'metadata' field :

In [ ]:
record = train_docs[0]

for key, value in record.items():
    print(f"{key}: {type(value)}")

metadata: <class 'dict'>
keywords: <class 'NoneType'>
sieverID: <class 'str'>
pagecount: <class 'str'>
content: <class 'str'>
tokenCount: <class 'str'>
images: <class 'list'>
tables: <class 'list'>


In [ ]:
metadata = train_docs[0]['metadata']

for key, value in metadata.items():
    print(f"{key}: {type(value)}")

gardian_id: <class 'str'>
source: <class 'str'>
url: <class 'str'>
description: <class 'str'>
id: <class 'str'>
title: <class 'str'>
language: <class 'str'>
resource_type: <class 'str'>
release_year: <class 'str'>
rights: <class 'str'>
dataNODE_id: <class 'str'>
keywords: <class 'list'>
geography: <class 'dict'>


In [ ]:
# Doing some random deep-dive

for r in range (2390,2400,1):

  metadata = train_docs[r]['metadata']
  for key, value in metadata.items():
    print(f"{key} : {value}")
  print(f'\n')

gardian_id : 4f948dfa8be4b3c1785f6784984a2c0a
source : gardian_index
url : http://www.alice.cnptia.embrapa.br:80/alice/bitstream/doc/942277/1/PROCI-2012.00194.pdf
description : he Brazilian government has announced, at COP 15, the country will target a reduction of its greenhouse gas emissions of between 36.1 and 38.9% from projected levels by 2020.
id : 1297191261
title : Estimates of enteric methane emission by steers bovine considering IPCC protocol in state of Para.
language : English
resource_type : Scientific Publication
release_year : 2023
rights : CC-BY-NC-ND
dataNODE_id : 3c96a085db75fa6a5841db7a63c41a7f
keywords : ['gase metano', 'pará', 'bovino', 'greenhouse gas emissions', 'methane emission', 'steers', 'state', 'para', 'government', 'reduction', 'levels']
geography : {'regions': ['South America', 'Latin America and the Caribbean', 'Americas', 'World'], 'countries': ['Brazil']}


gardian_id : 02673104ab539309b29c643345c7655a
source : gardian_index
url : http://www.alice.cnpt

In [ ]:
# Let's add some style to the deep-dive

for i in range(7):

    print(f"\n========== DOCUMENT {i} ==========")
    print(train_docs[i]["metadata"])


========== DOCUMENT 0 ==========
{'gardian_id': 'a5bf57d530bc56223e9e17b3f8aa149d', 'source': 'gardian_index', 'url': 'https://cgspace.cgiar.org/rest/bitstreams/9aa6310a-7b0a-419b-8099-e3807b25c0ad/retrieve', 'description': 'Climate information is now an agricultural input just like seeds, fertilizers or equipment which are at the basis of production. In a nutshell, this is what emerges from this assessment report on the CGIAR Program on Climate Change, Agriculture and Food Security (CCAFS) on the transmission of climate information (CI) and relevant agricultural advice to farmers in Kaffrine, Diourbel, Thies, Louga and Fatick administrative regions. Through this project, CCAFS seeks to contribute to improving the resilience of rural people to climate change by mainstreaming climate information more adequately in the planning and implementation of development activities.\nThanks to the involvement of the network of community rural radio stations (URAC), CCAFS estimates that two millio

In [ ]:
# Exploring the keys in the 'metadata' feature

all_metadata_keys = set()

for i in range(1000):
  all_metadata_keys.update(train_docs[i]["metadata"].keys())

print(all_metadata_keys)

{'release_year', 'keywords', 'geography', 'id', 'url', 'rights', 'title', 'resource_type', 'gardian_id', 'language', 'description', 'source', 'dataNODE_id'}


In [ ]:
# Check for keywords

none_count = 0

for i in range(len(train_docs)):
    if train_docs[i]["keywords"] is None:
        none_count += 1

print(f"Missing keywords: {none_count} / {len(train_docs)}")
print(f"Missing percentage: {none_count / len(train_docs) * 100:.2f}%")

Missing keywords: 68745 / 68745
Missing percentage: 100.00%


Note : Keywords can be used to complement the usefulness of the metadata feature in ***creating and curating*** our custom metadata.

### **Dataset Curation**

Data Curation will undergo three processes :

- Language Filtering : Here, filtering will result into only English texts.
- Geography Filtering : Here, the entire CGIAR corpus will be filtered into the region of Sub-Saharan Africa, only.
- Feature-Taxonomy & -Engineering : Features of the metadata will be further streamlined and made more SLM-custom.
- Adapt the customized order of the document arrangement in the base dataset onto our CGIAR-refined dataset

In [ ]:
# Inspecting language values

from collections import Counter

language_counts = Counter()

for record in train_docs:
  language = record['metadata'].get("language")
  language_counts[language] += 1

print(language_counts)

Counter({'English': 48325, 'Portuguese': 11463, 'French': 6274, 'Spanish': 1505, 'Indonesian': 251, 'Spanish, Castilian': 235, 'Hindi': 192, 'Vietnamese': 102, 'Russian': 66, 'Arabic': 55, 'Amharic': 42, 'Chinese': 34, '': 27, 'Swahili': 26, 'German': 17, 'Burmese': 16, 'Bengali': 12, 'Central Khmer': 12, 'Thai': 11, 'Tajik': 9, 'Uzbek': 6, 'Lao': 6, 'Italian': 6, 'Romanian, Moldavian, Moldovan': 4, 'Kinyarwanda': 3, 'Japanese': 3, 'Korean': 3, 'Nepali': 3, 'Ndonga': 2, 'Malay': 2, 'Shona': 2, 'Dinka': 2, 'Oromo': 2, 'Tagalog': 2, 'Bari': 2, 'Shilluk': 2, 'Oriya': 2, 'Dutch': 1, 'Yoruba': 1, 'Chichewa, Chewa, Nyanja': 1, 'Turkmen': 1, 'Georgian': 1, 'Otuho (Lotuko)': 1, 'Bambara': 1, 'Teso (Ateso)': 1, 'Urdu': 1, 'Luo (Dholuo)': 1, 'Moru': 1, 'Xhosa': 1, 'Nuer': 1, 'Ganda': 1, 'Telugu': 1, 'Tetum': 1, 'Malagasy': 1, 'Zande': 1, 'Sinhala, Sinhalese': 1})


Note : We now clearly know there are 48,325 English documents, which are subject to our SLM focus.

In [ ]:
# Inspecting geography values

# During 'data inspection,' we could see that geography is a 'dictionary' feature
# feature the 'metadata' feature. Let's open the can a little and take a peep!

geo_counts = Counter()

for record in train_docs:
  geo_regions = record['metadata'].get("geography").get("regions")
  for geo in geo_regions:
    geo_counts[geo]
  geo_counts[geo] += 1

print(geo_counts)

Counter({'World': 43666, 'Americas': 8101, 'Africa': 3208, 'Asia': 2530, 'Northern America': 2417, 'Europe': 1958, 'Southern Africa': 1002, 'Eastern Africa': 845, 'Western Asia': 692, 'Western Africa': 681, 'South-eastern Asia': 619, 'Northern Africa': 394, 'Oceania': 375, 'Southern Asia': 283, 'Central America': 279, 'Middle Africa': 277, 'Sub-Saharan Africa': 242, 'Latin America and the Caribbean': 198, 'Eastern Asia': 177, 'Northern Europe': 157, 'Caribbean': 142, 'South America': 128, 'Central Asia': 119, 'Eastern Europe': 84, 'Western Europe': 78, 'Southern Europe': 44, 'Micronesia': 22, 'Melanesia': 10, 'Polynesia': 8, 'Australia and New Zealand': 5, 'World Ocean': 4})


In [ ]:
# Peeping (randomly) into the 'geography' feature

for i in range(500,510,1):
  print (train_docs[i]['metadata']['geography'])

{'regions': ['Central America', 'Latin America and the Caribbean', 'Americas', 'World'], 'countries': ['Honduras']}
{'regions': [], 'countries': []}
{'regions': ['Eastern Africa', 'Sub-Saharan Africa', 'Africa', 'World', 'Southern Africa'], 'countries': ['Zambia']}
{'regions': ['Western Africa', 'Sub-Saharan Africa', 'Africa', 'World'], 'countries': ['Mali']}
{'regions': ['South America', 'Latin America and the Caribbean', 'Americas', 'World', 'Western Africa', 'Sub-Saharan Africa', 'Africa'], 'countries': ['Mali', 'Brazil']}
{'regions': ['South-eastern Asia', 'Asia', 'World', 'Eastern Asia', 'Northern America', 'Americas'], 'countries': ['United States of America', 'Thailand', 'China', 'Indonesia']}
{'regions': [], 'countries': []}
{'regions': ['Southern Asia', 'Asia', 'World', 'South America', 'Latin America and the Caribbean', 'Americas'], 'countries': ['Chile', 'Bangladesh']}
{'regions': ['South America', 'Latin America and the Caribbean', 'Americas', 'World'], 'countries': ['Colom

Note : Our SLM-focus is 'Sub-Saharan Africa'. We can see 'Sub-Saharan Africa' mingled (linked) with other regions in the 'regions' list. We also want such associations in our final corpus too.

In [ ]:
# Checking for amount of documents linked to 'Sub-Saharan Africa'

def is_sub_saharan_africa(doc):

  metadata = doc["metadata"]
  geography = metadata.get("geography", {})
  regions = geography.get("regions", [])

  return "Sub-Saharan Africa" in regions

sub_saharan_count = sum(
    is_sub_saharan_africa(doc) for doc in train_docs
)

print (f"Total CGIAR documents: {len(train_docs)}")
print (f"Number of documents linked to Sub-Saharan Africa: {sub_saharan_count}")


Total CGIAR documents: 68745
Number of documents linked to Sub-Saharan Africa: 18180


Note : We have 48,325 ***'English'*** documents and 18,180 ***'Sub-Saharan-Africa'*** documents. This gives us a clue that our final 'CGIAR' corpus will be less than 18,180 since we will have to eliminate non-'Sub-Saharan-Africa' documents whether they are in English or not.

Creating 'language and geography' filter :

In [ ]:
def is_sub_saharan_English(record):

  '''
    This function goes through a document and has two tests that such a document
    has to satisfy
  '''

  metadata = record["metadata"]

  # Test One(1): Language filter
  language = metadata.get("language")

  is_English = (language == "English")


  # Test Two (2): Geography filter
  geography = metadata.get("geography", {})
  regions = geography.get("regions", [])

  is_ssa = ("Sub-Saharan Africa" in regions)


  return is_English and is_ssa

Creating the newly-filtered dataset :

In [ ]:
sub_saharan_English_docs = [
    record
    for record in train_docs
    if is_sub_saharan_English(record)
]

print (f"Original length of CGIAR Corpus : {len(train_docs)}")
print (f"Number of documents linked to Sub-Saharan Africa: {len(sub_saharan_English_docs)}")

Original length of CGIAR Corpus : 68745
Number of documents linked to Sub-Saharan Africa: 14943


In [ ]:
# Inspecting a few of the documents in newly-filtered dataset

for i in range(3090,3100,1):
  record = sub_saharan_English_docs[i]

  print ('\n' + "="*80)
  print ("Title:", record['metadata'].get("title"))
  print ("Language:", record['metadata'].get("language"))
  print ("Countries:", record['metadata'].get("geography",{}).get("countries"))
  print ("Regions:", record['metadata'].get('geography',{}).get('regions'))



Title: The Status of Micronutrient and Sulphur in Some Plantation Crops of Different Ages in an Alfisol in Southern Nigeria
Language: English
Countries: ['Nigeria']
Regions: ['Western Africa', 'Sub-Saharan Africa', 'Africa', 'World']

Title: Fostering gender responsiveness and inclusivity in Nigeria’s agricultural seed policies: proposed reforms
Language: English
Countries: ['Nigeria']
Regions: ['Western Africa', 'Sub-Saharan Africa', 'Africa', 'World']

Title: Drought prediction in inland valleys rice-based production systems in the Sudan-Sahel Zone
Language: English
Countries: ['Sudan', 'Nigeria', 'Mali', 'Burkina Faso']
Regions: ['Western Africa', 'Sub-Saharan Africa', 'Africa', 'World', 'Northern Africa']

Title: FAC Working Paper 56. Agrarian Labour Relations in Zimbabwe after Over a Decade of Land and Agrarian Reform
Language: English
Countries: ['Zimbabwe']
Regions: ['Eastern Africa', 'Sub-Saharan Africa', 'Africa', 'World']

Title: Report on COHESA-Quadripartite joint training

SLM-Focus Feature-Taxonomy

In [ ]:
# Creating custom SLM-dataset columns

cgiar_docs_columns = ["document_id", "title", "topic", "crop", "agro_zone",
                      "text", "origin", "source_url", "license"]

In [ ]:
# Defining our custom dataset vocabulary for specific columns (especially columns)

TOPICS = ["crop_diseases", "pests", "soil_health", "climate_adaptation",
          "water_management", "fertiliser", "livestock", "post_harvest"]

CROPS = ["maize", "beans", "cassava","sorghum","groundnuts","livestock",
         "general"]

AGRO_ZONES = ["sub_humid", "semi_arid", "highland"]

Engineering the 'SLM-Focus' CGIAR features

In [ ]:
# Revisiting features again. Let's visit first row!

print (sub_saharan_English_docs[0])

{'metadata': {'gardian_id': 'a5bf57d530bc56223e9e17b3f8aa149d', 'source': 'gardian_index', 'url': 'https://cgspace.cgiar.org/rest/bitstreams/9aa6310a-7b0a-419b-8099-e3807b25c0ad/retrieve', 'description': 'Climate information is now an agricultural input just like seeds, fertilizers or equipment which are at the basis of production. In a nutshell, this is what emerges from this assessment report on the CGIAR Program on Climate Change, Agriculture and Food Security (CCAFS) on the transmission of climate information (CI) and relevant agricultural advice to farmers in Kaffrine, Diourbel, Thies, Louga and Fatick administrative regions. Through this project, CCAFS seeks to contribute to improving the resilience of rural people to climate change by mainstreaming climate information more adequately in the planning and implementation of development activities.\nThanks to the involvement of the network of community rural radio stations (URAC), CCAFS estimates that two million people have had acc

In [ ]:
# Combining the useful elements from the 'metadata' and 'content' features of
# the CGIAR dataset

def get_search_text(record):

  metadata = record["metadata"]

  title = metadata.get("title") or ""
  description = metadata.get("description") or ""
  keywords = metadata.get("keywords") or []
  content = record.get("content") or ""

  if isinstance(keywords, list):
    keywords_text = " ".join(str(k) for k in keywords)

  else:
    keywords_text = str(keywords)

  return " ".join([
      title,
      description,
      keywords_text,
      content
  ]).lower()

Derive more 'SLM-Features' from the metadata of the CGIAR dataset for our SLM Taxonomy

In [ ]:
# Derive for 'crop'

import re

def detect_crop(text):

  '''
      This function helps to categorize input texts based on the presence of specific crop-related terms,
      aiming to assign a single, most relevant crop, or a 'general' tag if multiple or no specific
      crop-types are detected.
  '''

  # Creating a dictionary of crop keywords and their relations in respect to the custom of our SLM focus
  crop_keywords = {
      "maize" : ["maize", "corn"],
      "beans" : ["beans", "bean"],
      "cassava" : ["cassava"],
      "sorghum" : ["sorghum"],
      "groundnuts" : ["groundnut", "ground nuts", "groundnuts"],
      "peanuts" : ["peanut", "peanuts", "pea"],
      "livestock" : ["livestock", "cattle", "cow", "goat", "goats", "sheep",
                     "poultry", "chicken", "pig", "pigs"]
      }

  # Use a set to store unique matches
  unique_matches = set()

  for crop, keywords in crop_keywords.items():

    for keyword in keywords:

      # Use regex for whole-word matching to avoid partial-word matching (e.g., 'pea' in 'peach')
      pattern = r'\b' + re.escape(keyword) + r'\b'
      if re.search(pattern, text, re.IGNORECASE):
        unique_matches.add(crop)

  if len(unique_matches) == 1:
    return list(unique_matches)[0] # Return the single unique crop

  elif len(unique_matches) > 1:
    return "general"

  else:
    return "general"

In [ ]:
def detect_topic(text):

  '''
      This function attempts to categorize input text into one of the predefined
      agricultural topics and its relations by checking how many of the topic's keywords appear
      in the text. The topic with the most keyword matches is considered the 'best' topic.
      If no keywords are found for any topic, it returns None
  '''

  topic_keywords = {
        "crop_diseases": ["disease", "diseases", "mosaic", "rust", "blight",
                          "wilt", "virus", "fungal"],

        "pests": ["pest", "pests", "fall armyworm", "aphid", "stem borer",
                  "insect", "weevil"],

        "soil_health": ["soil health", "soil acidity", "soil fertility",
                        "erosion", "cover crop", "compost"],

        "climate_adaptation": ["climate change", "climate adaptation",
                               "climate resilience", "drought", "seasonal forecast",
                               "climate information", "heat stress"],

        "water_management": ["irrigation","rainwater", "water management",
                             "water harvesting", "drip irrigation",
                             "water conservation"],

        "fertiliser": ["fertilizer","fertiliser", "nitrogen", "phosphorus",
                       "potassium", "npk", "manure", "nutrient"],

        "livestock": ["livestock", "cattle", "goat", "sheep", "poultry", "chicken",
                      "fodder", "animal health"],

        "post_harvest": ["post harvest", "post-harvest", "storage", "stored grain",
                         "aflatoxin", "drying", "hermetic"]
    }


  # Creating a dictionary to store mapped word and amount of occurence in searched text
  scores = {}

  for topic, words in topic_keywords.items():

    topic_score = 0

    for word in words:

      # Ensure whole word matching for each keyword
      pattern = r'\b' + re.escape(word) + r'\b'
      if re.search(pattern, text, re.IGNORECASE):
        topic_score += 1
    scores[topic] = topic_score

  best_topic = max(scores, key=scores.get)

  # If no keywords are found for any topic, it returns "general".
  if scores[best_topic] == 0:
        return "general"

  return best_topic

In [ ]:
# Derive for agro-zone

import re

def detect_agro_zone(text):

    '''
      It's a simple rule-based classifier that looks for direct mentions of specific geographical or
      climatic descriptors to assign an agro-zone category to the text.
  '''

    agro_zone_keywords = {

        "highland": ["highland", "highlands", "high land", "high lands", "upland",
                     "uplands", "mountain", "mountains", "montane"],

        "semi_arid": ["semi-arid", "semi arid", "semiarid", "arid", "arid zone",
                      "arid region", "arid areas"],

        "sub_humid": ["sub-humid", "sub humid", "subhumid", "humid", "humid zone",
                      "humid region", "humid areas"],
    }

    detected_zones = []

    for zone, keywords in agro_zone_keywords.items():

        for keyword in keywords:

            pattern = r'\b' + re.escape(keyword) + r'\b'

            if re.search(pattern, text, re.IGNORECASE):
                detected_zones.append(zone)
                break

    # Convert the list into a single string
    if detected_zones:
        return ", ".join(detected_zones)

    # Return this when no zone is detected
    return None

Combining all functions to get desired SLM dataset 'transform' functions

In [ ]:
def transform_cgiar_record(record):

    metadata = record["metadata"]

    text_for_classification = get_search_text(record)

    return {
        "document_id": metadata.get("gardian_id"),                # direct
        "title": metadata.get("title"),                           # direct
        "topic": detect_topic(text_for_classification),           # derived
        "crop": detect_crop(text_for_classification),             # derived
        "agro_zone": detect_agro_zone(text_for_classification),   # derived
        "text": record.get("content"),                            # direct
        "origin": "cgiar",                                        # direct
        "source_url": metadata.get("url"),                        # direct
        "license": metadata.get("rights")                         # direct
    }

In [ ]:
# Let's test and examine first five-hundred rows of our newly-formed dataset

import pandas as pd

examples = [
    transform_cgiar_record(sub_saharan_English_docs[i])
    for i in range(500)
]

cgiar_examples = pd.DataFrame(examples)

cgiar_examples

,document_id,title,topic,crop,agro_zone,text,origin,source_url,license
0,a5bf57d530bc56223e9e17b3f8aa149d,Impact assessment of communicating seasonal cl...,climate_adaptation,general,None,Climate information is now an agricultural inp...,cgiar,https://cgspace.cgiar.org/rest/bitstreams/9aa6...,N/A
1,b2d98cf83000b3953e0b2e42193cb2c7,Relations between Somali Regional State and So...,general,general,None,This research memo provides a snapshot analysi...,cgiar,,OGL-UK
2,30704b71b2d808c00a59b53919bd3d4b,Desertification and Climate Change in Africa,climate_adaptation,general,None,Desertification has increased in African dryla...,cgiar,https://cgspace.cgiar.org/rest/bitstreams/2594...,CP
3,07ee12b1ffd4ceadd24963e16b099818,CPWF Volta V5 Main Message Poster,general,general,None,A partner of the\nVolta Basin Development Chal...,cgiar,https://cgspace.cgiar.org/rest/bitstreams/ea5f...,N/A
4,a5b9ef90d76e5bc5295fb1932577348d,SC8 Share Fair Posters: Sustaining Food Availa...,livestock,general,None,"Barbara Rischkowsky, ICARDA\nB.Rischkowsky@cg...",cgiar,https://cgspace.cgiar.org/rest/bitstreams/01b7...,CC-BY
...,...,...,...,...,...,...,...,...,...
495,7bc71494692903a24ff7573dcb0b1dc9,Regional markets outlook and lessons for the d...,livestock,general,None,Regional Markets Outlook and Lessons for \nthe...,cgiar,https://cgspace.cgiar.org/rest/bitstreams/7d5f...,CC-BY
496,e32eaab00792f8b016f80319da269891,Investing in resilience: Climate risks and opp...,climate_adaptation,general,None,Investing in Resilience\nClimate Risks & Oppor...,cgiar,https://cgspace.cgiar.org/rest/bitstreams/7c0f...,CC-BY
497,aa5a01c5554dcbac309e123b69170721,Bridging the gap: decomposing sources of gende...,general,groundnuts,None,Female plot managers in Sub-Saharan Africa oft...,cgiar,http://oar.icrisat.org/11338/1/2019_GTD%20Mugi...,N/A
498,c7c0695e26b7eea74d206699e18aafbb,Evidence that rodent control strategies ought ...,post_harvest,general,None,Rodents pose serious threats to human health a...,cgiar,,N/A


In [ ]:
# Number of missing values in each column

missing_percentage = (
    cgiar_examples.isnull().sum() / len(cgiar_examples) * 100
).round(2)

print(missing_percentage)

document_id     0.0
title           0.0
topic           0.0
crop            0.0
agro_zone      78.8
text            0.0
origin          0.0
source_url      0.0
license         0.0
dtype: float64


We can take care of the missing values by refining the agro-zone detection function via expanding its classification conditions. Extended environmental specifications and evidences and crop presence-prevalence can be used to extend classification conditions in order to reduce the amount of missing values present.

In [ ]:
# Creating new agro-zone detection function

# Enriching the environmental specifications and evidences and crop presence-prevalence.

import re
import random

def detect_agroZone(text):

    # =========================================================
    # 1. EXPLICIT AGRO-ZONE EVIDENCE
    # =========================================================

    explicit_zone_keywords = {
        "highland": ["highland", "highlands", "high land", "high lands", "upland",
                     "uplands", "mountain", "mountains", "montane"],

        "semi_arid": ["semi-arid", "semi arid", "semiarid", "arid", "arid zone",
                      "arid region", "arid areas"],

        "sub_humid": ["sub-humid", "sub humid", "subhumid", "humid", "humid zone",
                      "humid region", "humid areas"],
    }


    explicit_zones = []

    for zone, keywords in explicit_zone_keywords.items():

        for keyword in keywords:

            pattern = r'\b' + re.escape(keyword) + r'\b'

            if re.search(pattern, text, re.IGNORECASE):
                explicit_zones.append(zone)
                break


    # =========================================================
    # 2. EXPLICIT EVIDENCE ALWAYS WINS
    # =========================================================

    if explicit_zones:
        return ", ".join(explicit_zones)

    # =========================================================
    # 3. ENVIRONMETAL EVIDENCE
    # =========================================================

    environmental_evidence = {

    "highland": [

        # Elevation / terrain
        "high elevation", "high elevations", "high altitude", "high altitudes",
        "elevated terrain", "elevated areas", "elevated region", "upland", "uplands",
        "upland area", "upland areas", "upland region", "upland regions", "mountainous",
        "mountainous area", "mountainous areas", "mountainous region", "mountainous regions",
        "mountain region", "mountain regions", "mountain slopes", "steep slopes",
        "steep terrain",

        # Temperature
        "cool temperatures", "cool temperature", "cool climate", "cooler temperatures",
        "low temperatures", "lower temperatures", "cold temperatures",

        # Highland-specific climatic description
        "highland climate", "highland environment"
    ],

    "semi_arid": [

        # Rainfall
        "low rainfall", "low precipitation", "limited rainfall", "limited precipitation",
        "low annual rainfall", "low annual precipitation", "erratic rainfall",
        "variable rainfall", "rainfall variability", "high rainfall variability",
        "unreliable rainfall", "unpredictable rainfall",

        # Growing / rainy season
        "short rainy season", "short wet season", "short growing season",
        "short growing period", "long dry season", "prolonged dry season",
        "extended dry season",

        # Water / moisture stress
        "water scarcity", "water stress", "water shortage", "moisture stress",
        "soil moisture stress",

        # Drought
        "drought", "drought conditions", "frequent drought", "frequent droughts",
        "recurrent drought",  "recurrent droughts", "prolonged drought", "prolonged droughts",
        "dry conditions", "dry climate"
    ],

    "sub_humid": [

        # Rainfall
        "moderate rainfall", "moderate precipitation", "adequate rainfall", "adequate precipitation",
        "reliable rainfall", "reliable precipitation", "seasonal rainfall","seasonal precipitation",

        # Growing / rainy season
        "long rainy season", "long wet season", "long growing season", "long growing period",
        "extended growing season",

        # Moisture / climate
        "moderate moisture", "moderate humidity", "moist climate", "moist conditions", "moist environment",

        # Rainfall range language
        "higher rainfall", "higher precipitation", "sufficient rainfall", "sufficient precipitation"
    ]
}

    environmental_scores = {
        "highland": 0,
        "semi_arid": 0,
        "sub_humid": 0,
    }



    for zone, terms in environmental_evidence.items():

        for term in terms:

            pattern = r'\b' + re.escape(term) + r'\b'

            if re.search(pattern, text, re.IGNORECASE):
                environmental_scores[zone] += 1


    # =========================================================
    # 3. CROP EVIDENCE
    # =========================================================

    crop_zone_mapping = {

    "maize": {
        "highland": 2, "semi_arid": 1, "sub_humid": 3
    },

    "sorghum": {
        "highland": 1, "semi_arid": 3, "sub_humid": 2
    },

    "groundnuts": {
        "highland": 1, "semi_arid": 3, "sub_humid": 2
    },

    "cassava": {
        "highland": 0, "semi_arid": 0, "sub_humid": 3
    },

    "beans": {
        "highland": 3, "semi_arid": 2, "sub_humid": 2
    }

        }


    crop_scores = {
        "highland": 0,
        "semi_arid": 0,
        "sub_humid": 0,
    }


    for crop, zone_scores in crop_zone_mapping.items():

        crop_pattern = r'\b' + re.escape(crop) + r'\b'

        if re.search(crop_pattern, text, re.IGNORECASE):

            for zone, score in zone_scores.items():
                crop_scores[zone] += score




    # =========================================================
    # 5. COMBINE CROP + ENVIRONMENTAL EVIDENCE
    # =========================================================

    combined_scores = {}

    for zone in crop_scores:

        combined_scores[zone] = (
            crop_scores[zone] +
            environmental_scores[zone]
        )


    # =========================================================
    # 6. FIND STRONGEST INFERRED ZONE
    # =========================================================

    max_score = max(combined_scores.values())

    if max_score > 0:

        strongest_zones = [
            zone
            for zone, score in combined_scores.items()
            if score == max_score
        ]

        # If there is a tie, choose one rather than None
        return random.choice(strongest_zones)


    # =========================================================
    # 7. ABSOLUTELY NO EVIDENCE
    # =========================================================

    return None

In [ ]:
# Combine newly-refined with all previous functions

def transform_cgiarRecord(record):

    metadata = record["metadata"]

    text_for_classification = get_search_text(record)

    return {
        "document_id": metadata.get("gardian_id"),                # direct
        "title": metadata.get("title"),                           # direct
        "topic": detect_topic(text_for_classification),           # derived
        "crop": detect_crop(text_for_classification),             # derived
        "agro_zone": detect_agroZone(text_for_classification),   # derived
        "text": record.get("content"),                            # direct
        "origin": "cgiar",                                        # direct
        "source_url": metadata.get("url"),                        # direct
        "license": metadata.get("rights")                         # direct
    }

In [ ]:
# Let's test and examine first five-hundred rows of our newly-formed dataset

import pandas as pd

examples = [
    transform_cgiarRecord(sub_saharan_English_docs[i])
    for i in range(500)
]

new_cgiar_examples = pd.DataFrame(examples)

new_cgiar_examples

,document_id,title,topic,crop,agro_zone,text,origin,source_url,license
0,a5bf57d530bc56223e9e17b3f8aa149d,Impact assessment of communicating seasonal cl...,climate_adaptation,general,None,Climate information is now an agricultural inp...,cgiar,https://cgspace.cgiar.org/rest/bitstreams/9aa6...,N/A
1,b2d98cf83000b3953e0b2e42193cb2c7,Relations between Somali Regional State and So...,general,general,None,This research memo provides a snapshot analysi...,cgiar,,OGL-UK
2,30704b71b2d808c00a59b53919bd3d4b,Desertification and Climate Change in Africa,climate_adaptation,general,None,Desertification has increased in African dryla...,cgiar,https://cgspace.cgiar.org/rest/bitstreams/2594...,CP
3,07ee12b1ffd4ceadd24963e16b099818,CPWF Volta V5 Main Message Poster,general,general,None,A partner of the\nVolta Basin Development Chal...,cgiar,https://cgspace.cgiar.org/rest/bitstreams/ea5f...,N/A
4,a5b9ef90d76e5bc5295fb1932577348d,SC8 Share Fair Posters: Sustaining Food Availa...,livestock,general,sub_humid,"Barbara Rischkowsky, ICARDA\nB.Rischkowsky@cg...",cgiar,https://cgspace.cgiar.org/rest/bitstreams/01b7...,CC-BY
...,...,...,...,...,...,...,...,...,...
495,7bc71494692903a24ff7573dcb0b1dc9,Regional markets outlook and lessons for the d...,livestock,general,sub_humid,Regional Markets Outlook and Lessons for \nthe...,cgiar,https://cgspace.cgiar.org/rest/bitstreams/7d5f...,CC-BY
496,e32eaab00792f8b016f80319da269891,Investing in resilience: Climate risks and opp...,climate_adaptation,general,sub_humid,Investing in Resilience\nClimate Risks & Oppor...,cgiar,https://cgspace.cgiar.org/rest/bitstreams/7c0f...,CC-BY
497,aa5a01c5554dcbac309e123b69170721,Bridging the gap: decomposing sources of gende...,general,groundnuts,None,Female plot managers in Sub-Saharan Africa oft...,cgiar,http://oar.icrisat.org/11338/1/2019_GTD%20Mugi...,N/A
498,c7c0695e26b7eea74d206699e18aafbb,Evidence that rodent control strategies ought ...,post_harvest,general,None,Rodents pose serious threats to human health a...,cgiar,,N/A


In [ ]:
# Number of missing values in each column

missing_percentage = (
    new_cgiar_examples.isnull().sum() / len(new_cgiar_examples) * 100
).round(2)

print(missing_percentage)

document_id     0.0
title           0.0
topic           0.0
crop            0.0
agro_zone      55.2
text            0.0
origin          0.0
source_url      0.0
license         0.0
dtype: float64


Now we can see a great percentage drop in missing values from 78.8% to 55.2%. Let's revisit our 'get_search_text' function in order to enrich it with more features from the 'metadata' or general features of whole dataset.

We can enrich it with geography : country and regions. These countries and regions can provide us with agro-zone mapping for our agroZone function

In [ ]:
# Combining the useful elements from the 'metadata' and 'content' features of
# the CGIAR dataset

# We bring in 'region' and 'country' features and use them as ways to extract
# and classify different agro-zones.

def get_search_Text(record):

  metadata = record["metadata"]

  title = metadata.get("title") or ""
  description = metadata.get("description") or ""
  keywords = metadata.get("keywords") or []
  content = record.get("content") or ""
  country = metadata.get("country") or ""
  region = metadata.get("region") or ""

  if isinstance(keywords, list):

    keywords_text = " ".join(str(k) for k in keywords)

  else:

    keywords_text = str(keywords)


  return " ".join([
      title,
      description,
      keywords_text,
      content,
      country,
      region,
  ]).lower()

In [ ]:
# Creating new agro-zone detection function

import re
import random

def detect_agro_Zone(text):

    # =========================================================
    # 1. EXPLICIT AGRO-ZONE EVIDENCE
    # =========================================================

    explicit_zone_keywords = {
        "highland": ["highland", "highlands", "high land", "high lands", "upland",
                     "uplands", "mountain", "mountains", "montane"],

        "semi_arid": ["semi-arid", "semi arid", "semiarid", "arid", "arid zone",
                      "arid region", "arid areas"],

        "sub_humid": ["sub-humid", "sub humid", "subhumid", "humid", "humid zone",
                      "humid region", "humid areas"],
    }


    explicit_zones = []

    for zone, keywords in explicit_zone_keywords.items():

        for keyword in keywords:

            pattern = r'\b' + re.escape(keyword) + r'\b'

            if re.search(pattern, text, re.IGNORECASE):
                explicit_zones.append(zone)
                break


    # =========================================================
    # 2. EXPLICIT EVIDENCE ALWAYS WINS
    # =========================================================

    if explicit_zones:
        return ", ".join(explicit_zones)

    # =========================================================
    # 3. COUNTRY-LEVEL AGRO-ZONE EVIDENCE
    # =========================================================

    country_zone_mapping = {

        # -----------------------------------------------------
        # WEST AFRICA
        # -----------------------------------------------------

        "senegal": {"highland": 0,"semi_arid": 3, "sub_humid": 1},

        "gambia": {"highland": 0, "semi_arid": 2,"sub_humid": 2},

        "guinea bissau": {"highland": 0, "semi_arid": 0, "sub_humid": 3},

        "guinea": {"highland": 2, "semi_arid": 1,"sub_humid": 3},

        "sierra leone": {"highland": 1,"semi_arid": 0, "sub_humid": 3},

        "liberia": { "highland": 1, "semi_arid": 0, "sub_humid": 3},

        "cote d'ivoire": {"highland": 1, "semi_arid": 0, "sub_humid": 3},

        "côte d'Ivoire": {"highland": 1, "semi_arid": 0, "sub_humid": 3},

        "ivory coast": {"highland": 1, "semi_arid": 0,"sub_humid": 3},

        "ghana": {"highland": 1, "semi_arid": 2, "sub_humid": 3},

        "togo": {"highland": 1, "semi_arid": 1, "sub_humid": 3},

        "benin": {"highland": 0, "semi_arid": 2, "sub_humid": 3},

        "mali": {
            "highland": 0,
            "semi_arid": 3,
            "sub_humid": 1
        },

        "niger": {
            "highland": 0,
            "semi_arid": 3,
            "sub_humid": 1
        },

        "nigeria": {
            "highland": 1,
            "semi_arid": 2,
            "sub_humid": 3
        },

        "mauritania": {
            "highland": 0,
            "semi_arid": 3,
            "sub_humid": 0
        },

        "cape verde": {
            "highland": 1,
            "semi_arid": 3,
            "sub_humid": 0
        },


        # -----------------------------------------------------
        # EAST AFRICA
        # -----------------------------------------------------

        "ethiopia": {
            "highland": 3,
            "semi_arid": 2,
            "sub_humid": 2
        },

        "eritrea": {
            "highland": 2,
            "semi_arid": 3,
            "sub_humid": 1
        },

        "djibouti": {
            "highland": 1,
            "semi_arid": 3,
            "sub_humid": 0
        },

        "somalia": {
            "highland": 0,
            "semi_arid": 3,
            "sub_humid": 0
        },

        "kenya": {
            "highland": 3,
            "semi_arid": 3,
            "sub_humid": 2
        },

        "uganda": {
            "highland": 2,
            "semi_arid": 1,
            "sub_humid": 3
        },

        "rwanda": {
            "highland": 3,
            "semi_arid": 0,
            "sub_humid": 3
        },

        "burundi": {
            "highland": 3,
            "semi_arid": 0,
            "sub_humid": 3
        },

        "tanzania": {
            "highland": 2,
            "semi_arid": 2,
            "sub_humid": 3
        },

        "south sudan": {
            "highland": 0,
            "semi_arid": 3,
            "sub_humid": 2
        },

        "sudan": {
            "highland": 0,
            "semi_arid": 3,
            "sub_humid": 1
        },


        # -----------------------------------------------------
        # CENTRAL AFRICA
        # -----------------------------------------------------

        "cameroon": {
            "highland": 2,
            "semi_arid": 1,
            "sub_humid": 3
        },

        "central african republic": {
            "highland": 0,
            "semi_arid": 1,
            "sub_humid": 3
        },

        "chad": {
            "highland": 0,
            "semi_arid": 3,
            "sub_humid": 1
        },

        "democratic republic of the congo": {
            "highland": 2,
            "semi_arid": 0,
            "sub_humid": 3
        },

        "drc": {
            "highland": 2,
            "semi_arid": 0,
            "sub_humid": 3
        },

        "republic of the congo": {
            "highland": 0,
            "semi_arid": 0,
            "sub_humid": 3
        },

        "gabon": {
            "highland": 1,
            "semi_arid": 0,
            "sub_humid": 3
        },

        "equatorial guinea": {
            "highland": 1,
            "semi_arid": 0, "sub_humid": 3
        },

        "são tomé and principe": {"highland": 2, "semi_arid": 1, "sub_humid": 3},

        "são tomé": {"highland": 2, "semi_arid": 1, "sub_humid": 3},

        "são tomé": {"highland": 2,"semi_arid": 1, "sub_humid": 3},

        "principe": {"highland": 2,"semi_arid": 0, "sub_humid": 3},


        # -----------------------------------------------------
        # SOUTHERN AFRICA
        # -----------------------------------------------------

        "angola": { "highland": 2, "semi_arid": 2, "sub_humid": 2
        },

        "zambia": {"highland": 1, "semi_arid": 2, "sub_humid": 3
        },

        "zimbabwe": {"highland": 2, "semi_arid": 3, "sub_humid": 1
        },

        "malawi": {"highland": 3, "semi_arid": 1, "sub_humid": 3
        },

        "mozambique": {"highland": 1,"semi_arid": 2, "sub_humid": 3},

        "namibia": {"highland": 1, "semi_arid": 3, "sub_humid": 0},

        "botswana": {"highland": 0,"semi_arid": 3, "sub_humid": 0},

        "south africa": { "highland": 2, "semi_arid": 3, "sub_humid": 2},

        "lesotho": {"highland": 3, "semi_arid": 2, "sub_humid": 1},

        "eswatini": {"highland": 2, "semi_arid": 1, "sub_humid": 3},


        # -----------------------------------------------------
        # INDIAN OCEAN
        # -----------------------------------------------------

        "madagascar": {"highland": 3, "semi_arid": 2, "sub_humid": 2},

        "comoros": {"highland": 2, "semi_arid": 0, "sub_humid": 3},

        "mauritius": {"highland": 2, "semi_arid": 0, "sub_humid": 3},

        "seychelles": {"highland": 1, "semi_arid": 0, "sub_humid": 3}
    }


    country_scores = {
        "highland": 0,
        "semi_arid": 0,
        "sub_humid": 0
    }


    for country, zone_scores in country_zone_mapping.items():

        pattern = r'\b' + re.escape(country) + r'\b'

        if re.search(pattern, text, re.IGNORECASE):

            for zone, score in zone_scores.items():

                # Country evidence is slightly weaker
                # than region-specific evidence
                country_scores[zone] += score


    # =========================================================
    # 4. REGION-LEVEL AGRO-ZONE EVIDENCE
    # =========================================================

    region_zone_mapping = {

        # -----------------------------------------------------
        # BROAD AFRICAN REGIONS
        # -----------------------------------------------------

        "sahel": {
            "highland": 0,
            "semi_arid": 3,
            "sub_humid": 1
        },

        "sahelian zone": {
            "highland": 0,
            "semi_arid": 3,
            "sub_humid": 1
        },

        "sudanian zone": {
            "highland": 0,
            "semi_arid": 2,
            "sub_humid": 3
        },

        "guinea savanna": {
            "highland": 0,
            "semi_arid": 1,
            "sub_humid": 3
        },

        "sudan savanna": {
            "highland": 0,
            "semi_arid": 3,
            "sub_humid": 2
        },

        "guinea savanna zone": {
            "highland": 0,
            "semi_arid": 1,
            "sub_humid": 3
        },

        "sudan savanna zone": {
            "highland": 0,
            "semi_arid": 3,
            "sub_humid": 2
        },

        "northern savanna": {
            "highland": 0,
            "semi_arid": 3,
            "sub_humid": 2
        },

        "forest zone": {
            "highland": 1,
            "semi_arid": 0,
            "sub_humid": 3
        },

        "forest belt": {
            "highland": 1,
            "semi_arid": 0,
            "sub_humid": 3
        },

        "coastal west africa": {
            "highland": 0,
            "semi_arid": 1,
            "sub_humid": 3
        },


        # -----------------------------------------------------
        # EAST AFRICA
        # -----------------------------------------------------

        "east africa": {
            "highland": 3,
            "semi_arid": 2,
            "sub_humid": 3
        },

        "east african highlands": {
            "highland": 3,
            "semi_arid": 1,
            "sub_humid": 3
        },

        "ethiopian highlands": {
            "highland": 3,
            "semi_arid": 2,
            "sub_humid": 2
        },

        "ethiopian plateau": {
            "highland": 3,
            "semi_arid": 2,
            "sub_humid": 2
        },

        "kenyan highlands": {
            "highland": 3,
            "semi_arid": 1,
            "sub_humid": 3
        },

        "great rift valley": {
            "highland": 2, "semi_arid": 3, "sub_humid": 2},

        "rift valley": {
            "highland": 2, "semi_arid": 3, "sub_humid": 2},

        "horn of africa": {"highland": 2,"semi_arid": 3,"sub_humid": 1},


        # -----------------------------------------------------
        # SOUTHERN AFRICA
        # -----------------------------------------------------

        "southern africa": {"highland": 2, "semi_arid": 3,"sub_humid": 2},

        "southern african highlands": {"highland": 3, "semi_arid": 2, "sub_humid": 2},

        "highveld": {"highland": 3, "semi_arid": 2, "sub_humid": 2},

        "lowveld": {"highland": 0, "semi_arid": 3,"sub_humid": 2},

        "zambezi basin": {"highland": 1, "semi_arid": 2, "sub_humid": 3},


        # -----------------------------------------------------
        # CENTRAL AFRICA
        # -----------------------------------------------------

        "central africa": {
            "highland": 1,
            "semi_arid": 0,
            "sub_humid": 3
        },

        "congo basin": {
            "highland": 0,
            "semi_arid": 0,
            "sub_humid": 3
        },

        "congo rainforest": {
            "highland": 0,
            "semi_arid": 0,
            "sub_humid": 3
        },

        "cameroon highlands": {
            "highland": 3,
            "semi_arid": 1,
            "sub_humid": 3
        },


        # -----------------------------------------------------
        # KENYA
        # -----------------------------------------------------

        "northern kenya": {
            "highland": 0,
            "semi_arid": 3,
            "sub_humid": 0
        },

        "eastern kenya": {
            "highland": 1,
            "semi_arid": 3,
            "sub_humid": 1
        },

        "western kenya": {
            "highland": 2,
            "semi_arid": 0,
            "sub_humid": 3
        },

        "central kenya": {
            "highland": 3,
            "semi_arid": 1,
            "sub_humid": 2
        },

        "coastal kenya": {
            "highland": 0,
            "semi_arid": 1,
            "sub_humid": 3
        },


        # -----------------------------------------------------
        # NIGERIA
        # -----------------------------------------------------

        "northern nigeria": {
            "highland": 0,
            "semi_arid": 3,
            "sub_humid": 2
        },

        "northwestern nigeria": {
            "highland": 0,
            "semi_arid": 3,
            "sub_humid": 1
        },

        "northeastern nigeria": {
            "highland": 0,
            "semi_arid": 3,
            "sub_humid": 1
        },

        "middle belt nigeria": {
            "highland": 1,
            "semi_arid": 1,
            "sub_humid": 3
        },

        "central nigeria": {
            "highland": 1,
            "semi_arid": 1,
            "sub_humid": 3
        },

        "southern nigeria": {
            "highland": 0,
            "semi_arid": 0,
            "sub_humid": 3
        },


        # -----------------------------------------------------
        # GHANA
        # -----------------------------------------------------

        "northern ghana": {
            "highland": 0,
            "semi_arid": 3,
            "sub_humid": 2
        },

        "upper east ghana": {
            "highland": 0,
            "semi_arid": 3,
            "sub_humid": 1
        },

        "upper west ghana": {
            "highland": 0,
            "semi_arid": 3,
            "sub_humid": 1
        },

        "middle ghana": {
            "highland": 1,
            "semi_arid": 1,
            "sub_humid": 3
        },

        "southern ghana": {
            "highland": 1,
            "semi_arid": 0,
            "sub_humid": 3
        },


        # -----------------------------------------------------
        # ETHIOPIA
        # -----------------------------------------------------

        "northern ethiopia": {
            "highland": 3,
            "semi_arid": 3,
            "sub_humid": 1
        },

        "central ethiopia": {
            "highland": 3,
            "semi_arid": 2,
            "sub_humid": 2
        },

        "eastern ethiopia": {
            "highland": 1,
            "semi_arid": 3,
            "sub_humid": 0
        },

        "western ethiopia": {
            "highland": 2,
            "semi_arid": 1,
            "sub_humid": 3
        },


        # -----------------------------------------------------
        # TANZANIA
        # -----------------------------------------------------

        "northern tanzania": {
            "highland": 2,
            "semi_arid": 2,
            "sub_humid": 3
        },

        "central tanzania": {
            "highland": 1,
            "semi_arid": 3,
            "sub_humid": 1
        },

        "southern tanzania": {
            "highland": 1,
            "semi_arid": 1,
            "sub_humid": 3
        },


        # -----------------------------------------------------
        # UGANDA
        # -----------------------------------------------------

        "northern uganda": {
            "highland": 0,
            "semi_arid": 2,
            "sub_humid": 3
        },

        "eastern uganda": {
            "highland": 1,
            "semi_arid": 1,
            "sub_humid": 3
        },

        "western uganda": {
            "highland": 3,
            "semi_arid": 0,
            "sub_humid": 3
        },


        # -----------------------------------------------------
        # ZIMBABWE
        # -----------------------------------------------------

        "northern zimbabwe": {
            "highland": 1,
            "semi_arid": 2,
            "sub_humid": 3
        },

        "southern zimbabwe": {
            "highland": 1,
            "semi_arid": 3,
            "sub_humid": 1
        },


        # -----------------------------------------------------
        # SOUTH AFRICA
        # -----------------------------------------------------

        "drakensberg": {
            "highland": 3,
            "semi_arid": 1,
            "sub_humid": 2
        }
    }


    region_scores = {
        "highland": 0,
        "semi_arid": 0,
        "sub_humid": 0
    }


    for region, zone_scores in region_zone_mapping.items():

        pattern = r'\b' + re.escape(region) + r'\b'

        if re.search(pattern, text, re.IGNORECASE):

            for zone, score in zone_scores.items():

                # Region evidence receives the full score
                region_scores[zone] += score

    # =========================================================
    # 3. ENVIRONMETAL EVIDENCE
    # =========================================================

    environmental_evidence = {

    "highland": [

        # Elevation / terrain
        "high elevation", "high elevations", "high altitude", "high altitudes",
        "elevated terrain", "elevated areas", "elevated region", "upland", "uplands",
        "upland area", "upland areas", "upland region", "upland regions", "mountainous",
        "mountainous area", "mountainous areas", "mountainous region", "mountainous regions",
        "mountain region", "mountain regions", "mountain slopes", "steep slopes",
        "steep terrain",

        # Temperature
        "cool temperatures", "cool temperature", "cool climate", "cooler temperatures",
        "low temperatures", "lower temperatures", "cold temperatures",

        # Highland-specific climatic description
        "highland climate", "highland environment"
    ],

    "semi_arid": [

        # Rainfall
        "low rainfall", "low precipitation", "limited rainfall", "limited precipitation",
        "low annual rainfall", "low annual precipitation", "erratic rainfall",
        "variable rainfall", "rainfall variability", "high rainfall variability",
        "unreliable rainfall", "unpredictable rainfall",

        # Growing / rainy season
        "short rainy season", "short wet season", "short growing season",
        "short growing period", "long dry season", "prolonged dry season",
        "extended dry season",

        # Water / moisture stress
        "water scarcity", "water stress", "water shortage", "moisture stress",
        "soil moisture stress",

        # Drought
        "drought", "drought conditions", "frequent drought", "frequent droughts",
        "recurrent drought",  "recurrent droughts", "prolonged drought", "prolonged droughts",
        "dry conditions", "dry climate"
    ],

    "sub_humid": [

        # Rainfall
        "moderate rainfall", "moderate precipitation", "adequate rainfall", "adequate precipitation",
        "reliable rainfall", "reliable precipitation", "seasonal rainfall","seasonal precipitation",

        # Growing / rainy season
        "long rainy season", "long wet season", "long growing season", "long growing period",
        "extended growing season",

        # Moisture / climate
        "moderate moisture", "moderate humidity", "moist climate", "moist conditions", "moist environment",

        # Rainfall range language
        "higher rainfall", "higher precipitation", "sufficient rainfall", "sufficient precipitation"
    ]
}

    environmental_scores = {
        "highland": 0,
        "semi_arid": 0,
        "sub_humid": 0,
    }



    for zone, terms in environmental_evidence.items():

        for term in terms:

            pattern = r'\b' + re.escape(term) + r'\b'

            if re.search(pattern, text, re.IGNORECASE):
                environmental_scores[zone] += 1


    # =========================================================
    # 3. CROP EVIDENCE
    # =========================================================

    crop_zone_mapping = {

    "maize": {
        "highland": 2, "semi_arid": 1, "sub_humid": 3
    },

    "sorghum": {
        "highland": 1, "semi_arid": 3, "sub_humid": 2
    },

    "groundnuts": {
        "highland": 1, "semi_arid": 3, "sub_humid": 2
    },

    "cassava": {
        "highland": 0, "semi_arid": 0, "sub_humid": 3
    },

    "beans": {
        "highland": 3, "semi_arid": 2, "sub_humid": 2
    }

        }


    crop_scores = {
        "highland": 0,
        "semi_arid": 0,
        "sub_humid": 0,
    }


    for crop, zone_scores in crop_zone_mapping.items():

        crop_pattern = r'\b' + re.escape(crop) + r'\b'

        if re.search(crop_pattern, text, re.IGNORECASE):

            for zone, score in zone_scores.items():
                crop_scores[zone] += score




    # =========================================================
    # 5. COMBINE CROP + ENVIRONMENTAL EVIDENCE
    # =========================================================

    combined_scores = {}

    for zone in crop_scores:

        combined_scores[zone] = (
            country_scores[zone]
            +
            region_scores[zone]
            +
            environmental_scores[zone]
            +
            crop_scores[zone]
        )


    # =========================================================
    # 6. FIND STRONGEST INFERRED ZONE
    # =========================================================

    max_score = max(combined_scores.values())

    if max_score > 0:

        strongest_zones = [
            zone
            for zone, score in combined_scores.items()
            if score == max_score
        ]

        # If there is a tie, choose one rather than None
        return random.choice(strongest_zones)


    # =========================================================
    # 7. ABSOLUTELY NO EVIDENCE
    # =========================================================

    return None

In [ ]:
# Combine with all previous functions

def transform_cgiar_Record(record):

    metadata = record["metadata"]

    text_for_classification = get_search_Text(record)

    return {
        "document_id": metadata.get("gardian_id"),                # direct
        "title": metadata.get("title"),                           # direct
        "topic": detect_topic(text_for_classification),           # derived
        "crop": detect_crop(text_for_classification),             # derived
        "agro_zone": detect_agro_Zone(text_for_classification),   # derived
        "text": record.get("content"),                            # direct
        "origin": "cgiar",                                        # direct
        "source_url": metadata.get("url"),                        # direct
        "license": metadata.get("rights")                         # direct
    }

In [ ]:
# Let's test and examine first five-hundred rows of our newly-formed dataset

import pandas as pd

examples = [
    transform_cgiar_Record(sub_saharan_English_docs[i])
    for i in range(500)
]

new_cgiarExamples = pd.DataFrame(examples)

new_cgiarExamples

,document_id,title,topic,crop,agro_zone,text,origin,source_url,license
0,a5bf57d530bc56223e9e17b3f8aa149d,Impact assessment of communicating seasonal cl...,climate_adaptation,general,semi_arid,Climate information is now an agricultural inp...,cgiar,https://cgspace.cgiar.org/rest/bitstreams/9aa6...,N/A
1,b2d98cf83000b3953e0b2e42193cb2c7,Relations between Somali Regional State and So...,general,general,highland,This research memo provides a snapshot analysi...,cgiar,,OGL-UK
2,30704b71b2d808c00a59b53919bd3d4b,Desertification and Climate Change in Africa,climate_adaptation,general,None,Desertification has increased in African dryla...,cgiar,https://cgspace.cgiar.org/rest/bitstreams/2594...,CP
3,07ee12b1ffd4ceadd24963e16b099818,CPWF Volta V5 Main Message Poster,general,general,None,A partner of the\nVolta Basin Development Chal...,cgiar,https://cgspace.cgiar.org/rest/bitstreams/ea5f...,N/A
4,a5b9ef90d76e5bc5295fb1932577348d,SC8 Share Fair Posters: Sustaining Food Availa...,livestock,general,sub_humid,"Barbara Rischkowsky, ICARDA\nB.Rischkowsky@cg...",cgiar,https://cgspace.cgiar.org/rest/bitstreams/01b7...,CC-BY
...,...,...,...,...,...,...,...,...,...
495,7bc71494692903a24ff7573dcb0b1dc9,Regional markets outlook and lessons for the d...,livestock,general,sub_humid,Regional Markets Outlook and Lessons for \nthe...,cgiar,https://cgspace.cgiar.org/rest/bitstreams/7d5f...,CC-BY
496,e32eaab00792f8b016f80319da269891,Investing in resilience: Climate risks and opp...,climate_adaptation,general,highland,Investing in Resilience\nClimate Risks & Oppor...,cgiar,https://cgspace.cgiar.org/rest/bitstreams/7c0f...,CC-BY
497,aa5a01c5554dcbac309e123b69170721,Bridging the gap: decomposing sources of gende...,general,groundnuts,sub_humid,Female plot managers in Sub-Saharan Africa oft...,cgiar,http://oar.icrisat.org/11338/1/2019_GTD%20Mugi...,N/A
498,c7c0695e26b7eea74d206699e18aafbb,Evidence that rodent control strategies ought ...,post_harvest,general,sub_humid,Rodents pose serious threats to human health a...,cgiar,,N/A


In [ ]:
# Number of missing values in each column

missing_percentage = (
    new_cgiarExamples.isnull().sum() / len(new_cgiarExamples) * 100
).round(2)

print(missing_percentage)

document_id     0.0
title           0.0
topic           0.0
crop            0.0
agro_zone      14.2
text            0.0
origin          0.0
source_url      0.0
license         0.0
dtype: float64


Reduction has seen great progress from 78.8% to 55.2% and now 14.2%. ✅

Let's inspect documents where missing values exists.

In [ ]:
missing_agro = new_cgiarExamples[
    new_cgiarExamples["agro_zone"].isna()
].copy()

print("Remaining missing:", len(missing_agro))

Remaining missing: 71


Let's also inspect information available to the classifier.

In [ ]:
missing_agro[
    ["document_id", "title", "text"]
].head(10)

,document_id,title,text
2,30704b71b2d808c00a59b53919bd3d4b,Desertification and Climate Change in Africa,Desertification has increased in African dryla...
3,07ee12b1ffd4ceadd24963e16b099818,CPWF Volta V5 Main Message Poster,A partner of the\nVolta Basin Development Chal...
17,0814f8cba43aa6d3ea8636cbb27f7ca2,Solar drying adds value to surplus fruit and v...,In Burkina Faso production from market gardeni...
19,8b741eba5f0e0a7f691fc6363b643ff2,Climate Research for Development (CR4D) Africa...,Reporting 2021 Policies #139\nClimate Research...
25,177ed3481982b5bac70f00385cade8b1,19TDaGRCTraitProfilingIB,Trait profiling of 35 collections D. alata acc...
28,7985e897ad3ab0ea0a028ea2f8b63405,Enterprise Survey 2010 (Bolivia),This research was conducted in Bolivia between...
39,b2c62a169bc3a42113ca443bb49c086d,Effects of ultrasonic combined with hydroxypro...,Foxtail millet is a specialty grain mainly con...
40,f97bd4e1824626deb4beb25f7b5fc66e,Integrating adaptation into REDD+: potential i...,REDD+ interventions can help both people and f...
45,531c0b63224615cc8bda66dbf94f1b87,Myeloperoxidase and other markers of neutrophi...,Introduction: Placental malaria (PM) is charac...
52,de7ccbce4808a10f489e6cd0090e9c41,Enterprise Survey 2009 (Hungary),The objective of the survey is to obtain feedb...


All information that can enrich our agro-zone mapping has been exhausted. Let's resign to assigning non-values as 'unknown.'

Let's apply this to whole dataset.


In [ ]:
# Applying refined functions to whole CGIAR-curated dataset

from tqdm.auto import tqdm                                        # progress bar

final_cgiar_docs = [
    transform_cgiar_Record(record)
    for record in tqdm(sub_saharan_English_docs,
                       desc="Transforming CGIAR documents")
]

Transforming CGIAR documents:   0%|          | 0/14943 [00:00<?, ?it/s]

In [ ]:
# Number of missing values in each column in

missing_percentage = (
    new_cgiarExamples.isnull().sum() / len(new_cgiarExamples) * 100
).round(2)

print(missing_percentage)

document_id     0.0
title           0.0
topic           0.0
crop            0.0
agro_zone      14.2
text            0.0
origin          0.0
source_url      0.0
license         0.0
dtype: float64


In [ ]:
# Accessing whole dataset matrix-shape

import pandas as pd

final_cgiar_docs = pd.DataFrame(final_cgiar_docs)
print (final_cgiar_docs.shape)
print('\n')
print(f"Five first rows of the refined CGIAR docoments :")

final_cgiar_docs.head()

(14943, 9)


Five first rows of the refined CGIAR docoments :


,document_id,title,topic,crop,agro_zone,text,origin,source_url,license
0,a5bf57d530bc56223e9e17b3f8aa149d,Impact assessment of communicating seasonal cl...,climate_adaptation,general,semi_arid,Climate information is now an agricultural inp...,cgiar,https://cgspace.cgiar.org/rest/bitstreams/9aa6...,N/A
1,b2d98cf83000b3953e0b2e42193cb2c7,Relations between Somali Regional State and So...,general,general,highland,This research memo provides a snapshot analysi...,cgiar,,OGL-UK
2,30704b71b2d808c00a59b53919bd3d4b,Desertification and Climate Change in Africa,climate_adaptation,general,None,Desertification has increased in African dryla...,cgiar,https://cgspace.cgiar.org/rest/bitstreams/2594...,CP
3,07ee12b1ffd4ceadd24963e16b099818,CPWF Volta V5 Main Message Poster,general,general,None,A partner of the\nVolta Basin Development Chal...,cgiar,https://cgspace.cgiar.org/rest/bitstreams/ea5f...,N/A
4,a5b9ef90d76e5bc5295fb1932577348d,SC8 Share Fair Posters: Sustaining Food Availa...,livestock,general,sub_humid,"Barbara Rischkowsky, ICARDA\nB.Rischkowsky@cg...",cgiar,https://cgspace.cgiar.org/rest/bitstreams/01b7...,CC-BY


In [ ]:
# Number of missing values in each column

missing_percentage = (
    final_cgiar_docs.isnull().sum() / len(final_cgiar_docs) * 100
).round(2)

print(missing_percentage)

document_id     0.00
title           0.00
topic           0.00
crop            0.00
agro_zone      13.94
text            0.00
origin          0.00
source_url      0.00
license         0.00
dtype: float64


In [ ]:
# Getting the amount of documents with missing values in 'agro_zone'

missing_agro = final_cgiar_docs[
    final_cgiar_docs["agro_zone"].isna()
].copy()

print("Remaining missing:", len(missing_agro))

Remaining missing: 2083


We have exhausted all available filters for the agro_zone. We can drop missing values.

The resultant data is sufficient for our SLM purpose.

In [ ]:
# Dropping all missing values since non-missing values in the agro_zone is
# sufficient for our SLM purpose.

final_cgiar_docs = final_cgiar_docs.dropna(subset=["agro_zone"])

In [ ]:
# Confirm whole dataset for missing values

final_cgiar_docs.isna().sum()

,0
document_id,0
title,0
topic,0
crop,0
agro_zone,0
text,0
origin,0
source_url,0
license,0


Next step is to concatenate both datasets (base dataset and the CGIAR-curated dataset). But before that, let's check for column-similarity and rectify if need be.

In [ ]:
# Check similarity of columns of both datasets

print (f"{base_df.columns}")
print(f"\n {final_cgiar_docs.columns}")

print(f"\n Do both datasets have same columns : {base_df.columns == final_cgiar_docs.columns}")

Index(['document_id', 'title', 'topic', 'crop', 'agro_zone', 'text', 'origin',
       'source_url', 'license'],
      dtype='object')

 Index(['document_id', 'title', 'topic', 'crop', 'agro_zone', 'text', 'origin',
       'source_url', 'license'],
      dtype='object')

 Do both datasets have same columns : [ True  True  True  True  True  True  True  True  True]


Note : There is something interesting in the way the contents of the first column of the base dataset (base_df[document_id]) was arranged. Let similarize that arrangements with the contents of the first column of the CGIAR-curated dataset (final_cgiar_docs[document_id]).

In [ ]:
# Since 'final_cgiar_docs' and 'cgiar_examples' share same first first five columns,
# let's with with 'cgiar_examples' for now

print("Base Dataset : \n")
display(base_df[['document_id', 'topic']].head())

print("\n\n CGIAR-Refined Dataset : \n")
display(cgiar_examples[['document_id', 'topic']].head())

Base Dataset : 



,document_id,topic
0,doc_dis_001,crop_diseases
1,doc_dis_002,crop_diseases
2,doc_dis_003,crop_diseases
3,doc_pes_001,pests
4,doc_pes_002,pests




 CGIAR-Refined Dataset : 



,document_id,topic
0,a5bf57d530bc56223e9e17b3f8aa149d,climate_adaptation
1,b2d98cf83000b3953e0b2e42193cb2c7,general
2,30704b71b2d808c00a59b53919bd3d4b,climate_adaptation
3,07ee12b1ffd4ceadd24963e16b099818,general
4,a5b9ef90d76e5bc5295fb1932577348d,livestock


We can see that in the document_id of the base dataset, the documents were named with some form of customized order where the middle term stood as prefix for the name of the topic of that data point, the numbers were arranged in accordance with the n-th position of that same topic and the first term stood as a constant name - doc.

Let's imitate this naming convention onto the CGIAR-curated dataset. To do, we have to concatenate both the base dataset and the CGIAR-curated dataset. This is to avoid having more than one datapoint sharing same 'document_id.'



Concatenating ***'Dataset One(1) : Base Dataset'*** and ***'Dataset Two(2) : CGIAR Source'*** :

In [ ]:
combined_df = pd.concat([base_df, final_cgiar_docs], axis = 0, ignore_index =True)

In [ ]:
combined_df.head()

,document_id,title,topic,crop,agro_zone,text,origin,source_url,license
0,doc_dis_001,Maize nitrogen deficiency symptoms,crop_diseases,maize,sub_humid,Nitrogen deficiency in maize shows as uniform ...,synthetic,NaN,CC0-1.0
1,doc_dis_002,Common bean rust management,crop_diseases,beans,highland,Bean rust appears as small reddish-brown pustu...,synthetic,NaN,CC0-1.0
2,doc_dis_003,Cassava mosaic disease,crop_diseases,cassava,sub_humid,"Cassava mosaic causes leaf distortion, chlorot...",synthetic,NaN,CC0-1.0
3,doc_pes_001,Fall armyworm in maize,pests,maize,semi_arid,Fall armyworm larvae feed in the whorl leaving...,synthetic,NaN,CC0-1.0
4,doc_pes_002,Stem borer in sorghum,pests,sorghum,semi_arid,Stem borers tunnel inside sorghum stalks causi...,synthetic,NaN,CC0-1.0


In [ ]:
combined_df.tail()

,document_id,title,topic,crop,agro_zone,text,origin,source_url,license
12879,e887cedd82b53781c699506ac9d8451f,Quarterly Labour Force Survey 2011: Second Qua...,general,general,semi_arid,The Quarterly Labour Force Survey (QLFS) is a ...,cgiar,,N/A
12880,53e617f9a24c7d50479dcf7aca07276e,Discussion Paper 23. Analysis of the curriculu...,general,general,sub_humid,\n M U S T E R \n Multi-Site Teacher Educati...,cgiar,https://assets.publishing.service.gov.uk/media...,OGL-UK
12881,b967fcb8620560dfa92ab126f519631b,Modeling the Impact of Climate Variability on ...,pests,general,sub_humid,Despite the functioning of tomato production i...,cgiar,https://cgspace.cgiar.org/rest/bitstreams/e6b9...,CC-BY-NC
12882,491fa3977027c70c2e8785bdae152572,CGIAR Initiative on Agroecology,livestock,livestock,semi_arid,Agroecology\nAgroecology is a transdisciplinar...,cgiar,https://cgspace.cgiar.org/rest/bitstreams/cd83...,CC-BY-NC
12883,56311e14cf132ee14cfc90a0cae656f6,Updating and estimating a Social Account Matri...,livestock,general,sub_humid,TMD DISCUSSION PAPER NO. 58 \n \n \n \n \nUpda...,cgiar,https://cgspace.cgiar.org/rest/bitstreams/b4b4...,N/A


Applying the ordered naming convention for whole concatenated dataset.

In [ ]:
# Define 'topic-document' prefix mapping

topic_prefix = {
    "crop_diseases" : "dis",
    "pests" : "pes",
    "soil_health" : "soi",
    "climate_adaptation" : "cli",
    "water_management" : "wat",
    "fertiliser" : "fer",
    "livestock" : "liv",
    "post_harvest" : "pos",
    "general" : "gen"
}


In [ ]:
# Generate IDs across the entire combined dataset

topic_counters = {}

new_document_ids = []

for index, row in combined_df.iterrows():

    topic = row['topic']

    # Get the prefix for this topic
    prefix_mapping = topic_prefix.get(topic, "gen")

    # Start the counter if this is the first occurence
    if topic not in topic_counters:
        topic_counters[topic] = 0

    # Increase the counter
    topic_counters[topic] += 1

    # Create the ID
    document_id = f'doc_{prefix_mapping}_{topic_counters[topic]:03d}'

    # Append the new ID to the list
    new_document_ids.append(document_id)

combined_df["document_id"] = new_document_ids


In [ ]:
# Confirming if the 'document_id' of both dataset now follow a uniform convention

combined_df.tail()

,document_id,title,topic,crop,agro_zone,text,origin,source_url,license
12879,doc_gen_3236,Quarterly Labour Force Survey 2011: Second Qua...,general,general,semi_arid,The Quarterly Labour Force Survey (QLFS) is a ...,cgiar,NaN,NaN
12880,doc_gen_3237,Discussion Paper 23. Analysis of the curriculu...,general,general,sub_humid,\n M U S T E R \n Multi-Site Teacher Educati...,cgiar,https://assets.publishing.service.gov.uk/media...,OGL-UK
12881,doc_pes_387,Modeling the Impact of Climate Variability on ...,pests,general,sub_humid,Despite the functioning of tomato production i...,cgiar,https://cgspace.cgiar.org/rest/bitstreams/e6b9...,CC-BY-NC
12882,doc_liv_2227,CGIAR Initiative on Agroecology,livestock,livestock,semi_arid,Agroecology\nAgroecology is a transdisciplinar...,cgiar,https://cgspace.cgiar.org/rest/bitstreams/cd83...,CC-BY-NC
12883,doc_liv_2228,Updating and estimating a Social Account Matri...,livestock,general,sub_humid,TMD DISCUSSION PAPER NO. 58 \n \n \n \n \nUpda...,cgiar,https://cgspace.cgiar.org/rest/bitstreams/b4b4...,NaN


Save to Drive

In [ ]:
from google.colab import drive

drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
combined_df.to_csv('/content/drive/MyDrive/combined_df.csv', index=False)

In [ ]:
import pandas as pd

combined_df = pd.read_csv('/content/drive/MyDrive/combined_df.csv')

In [ ]:
combined_df.shape

(12884, 9)

In [ ]:
# Take a 20% random sample of the combined_df
sampled_df = combined_df.sample(frac=0.2, random_state=42) # Using random_state for reproducibility

print(f"Original combined_df shape: {combined_df.shape}")
print(f"Sampled combined_df shape (20%): {sampled_df.shape}")

display(sampled_df.head())

Original combined_df shape: (12884, 9)
Sampled combined_df shape (20%): (2577, 9)


,document_id,title,topic,crop,agro_zone,text,origin,source_url,license
5268,doc_fer_463,Combining ability and performance of extra-ear...,fertiliser,general,semi_arid,CROPS AND SOILS RESEARCH PAPER\nCombining abil...,cgiar,https://cgspace.cgiar.org/rest/bitstreams/6b19...,CC-BY
9159,doc_dis_1679,Phytogeographical and sociolinguistical patter...,crop_diseases,general,sub_humid,BACKGROUND: Many fungal species in tropical Af...,cgiar,NaN,NaN
8,doc_soi_003,Compost quality for smallholders,soil_health,general,semi_arid,"Well-made compost should be dark, crumbly, and...",synthetic,NaN,CC0-1.0
7460,doc_cli_1245,CCAFS in West Africa,climate_adaptation,general,semi_arid,CCAFS IN \nWEST \nAFRICA\nClimate change has e...,cgiar,https://cgspace.cgiar.org/rest/bitstreams/ef34...,NaN
1579,doc_gen_389,Market institutions: enhancing the value of ru...,general,general,highland,\nDiscussion Paper 154 \n \n \n \n \n \n \nme...,cgiar,https://cgspace.cgiar.org/rest/bitstreams/674d...,NaN


In [ ]:
# Calculate the number of rows for the first 20%
num_rows_20_percent = int(len(combined_df) * 0.2)

# Select the first 20% of the DataFrame
first_20_percent_df = combined_df.iloc[:num_rows_20_percent]

print(f"Original combined_df shape: {combined_df.shape}")
print(f"First 20% of combined_df shape: {first_20_percent_df.shape}")

display(first_20_percent_df.head())

Original combined_df shape: (12884, 9)
First 20% of combined_df shape: (2576, 9)


,document_id,title,topic,crop,agro_zone,text,origin,source_url,license
0,doc_dis_001,Maize nitrogen deficiency symptoms,crop_diseases,maize,sub_humid,Nitrogen deficiency in maize shows as uniform ...,synthetic,NaN,CC0-1.0
1,doc_dis_002,Common bean rust management,crop_diseases,beans,highland,Bean rust appears as small reddish-brown pustu...,synthetic,NaN,CC0-1.0
2,doc_dis_003,Cassava mosaic disease,crop_diseases,cassava,sub_humid,"Cassava mosaic causes leaf distortion, chlorot...",synthetic,NaN,CC0-1.0
3,doc_pes_001,Fall armyworm in maize,pests,maize,semi_arid,Fall armyworm larvae feed in the whorl leaving...,synthetic,NaN,CC0-1.0
4,doc_pes_002,Stem borer in sorghum,pests,sorghum,semi_arid,Stem borers tunnel inside sorghum stalks causi...,synthetic,NaN,CC0-1.0


In [ ]:
# Save the randomly sampled 20% DataFrame
sampled_df.to_csv('/content/drive/MyDrive/sampled_combined_df_20_percent.csv', index=False)
print('Saved sampled_combined_df_20_percent.csv to Google Drive')

Saved sampled_combined_df_20_percent.csv to Google Drive


In [ ]:
# Save the first 20% DataFrame
first_20_percent_df.to_csv('/content/drive/MyDrive/first_20_percent_combined_df.csv', index=False)
print('Saved first_20_percent_combined_df.csv to Google Drive')

Saved first_20_percent_combined_df.csv to Google Drive


***Dataset Three(3) : The FAO Source***

In [ ]:
!pip install pdfplumber requests beautifulsoup4 -q

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 43.7/43.7 kB 3.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 66.5/66.5 kB 6.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 60.0/60.0 kB 4.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.6/6.6 MB 86.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.9/6.9 MB 128.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.7/3.7 MB 112.2 MB/s eta 0:00:00


In [ ]:
fao_sources = [
    {
        "url": "https://www.fao.org/4/i2230e/i2230e.pdf",
        "title": "Climate Change and Food Systems Resilience in Sub-Saharan Africa",
        "topic": "climate_adaptation",
        "crop": "multiple",
        "agro_zone": "sub_humid",
    },
    {
        "url": "https://www.fao.org/4/i2575e/i2575e00.pdf",
        "title": "Climate-Smart Agriculture Smallholder Adoption SSA",
        "topic": "climate_adaptation",
        "crop": "multiple",
        "agro_zone": "semi_arid",
    },
    {
        "url": "https://www.fao.org/4/i2230e/i2230e07.pdf",
        "title": "Climate Change Implications for Agriculture in Sub-Saharan Africa",
        "topic": "climate_adaptation",
        "crop": "multiple",
        "agro_zone": "semi_arid",
    },
    {
        "url": "https://www.fao.org/4/an112e/an112e00.pdf",
        "title": "Identifying Opportunities for CSA Investments in Africa",
        "topic": "food_security",
        "crop": "multiple",
        "agro_zone": "sub_humid",
    },
]

In [ ]:
import requests
import pdfplumber
import io
import re

def download_and_extract(url):
    response = requests.get(url, timeout=30)
    if response.status_code != 200:
        print(f"Failed: {url}")
        return ""
    with pdfplumber.open(io.BytesIO(response.content)) as pdf:
        pages = [page.extract_text() or "" for page in pdf.pages]
    return "\n".join(pages)

def clean_text(text):
    text = re.sub(r'\s+', ' ', text)               # collapse whitespace
    text = re.sub(r'[^\w\s.,;:()\-\/]', '', text)  # remove noise
    text = text.strip()
    return text

raw_texts = []
for doc in fao_sources:
    print(f"Downloading: {doc['title'][:50]}...")
    raw = download_and_extract(doc["url"])
    cleaned = clean_text(raw)
    raw_texts.append(cleaned)
    print(f"  → {len(cleaned)} chars extracted")

Downloading: Climate Change and Food Systems Resilience in Sub-...


  → 792780 chars extracted
Downloading: Climate-Smart Agriculture Smallholder Adoption SSA...
  → 94052 chars extracted
Downloading: Climate Change Implications for Agriculture in Sub...


  → 29132 chars extracted
Downloading: Identifying Opportunities for CSA Investments in A...
  → 356136 chars extracted


In [ ]:
!pip install langchain-text-splitters -q

In [ ]:
import pandas as pd

from langchain_text_splitters import RecursiveCharacterTextSplitter

splitter = RecursiveCharacterTextSplitter(
    chunk_size=400,      # small — good for SLM
    chunk_overlap=50,
    separators=["\n\n", "\n", ".", " "]
)

rows = []
for i, (doc, text) in enumerate(zip(fao_sources, raw_texts)):
    chunks = splitter.split_text(text)
    for j, chunk in enumerate(chunks):
        if len(chunk.strip()) < 60:
            continue
        rows.append({
            "document_id": f"fao_{i+1:03d}_{j+1:04d}",
            "title":       doc["title"],
            "topic":       doc["topic"],
            "crop":        doc["crop"],
            "agro_zone":   doc["agro_zone"],
            "text":        chunk.strip(),
            "origin":      "fao",
            "source_url":  doc["url"],
            "license":     "CC-BY-NC-SA-3.0-IGO",
        })

fao_df = pd.DataFrame(rows)
print(f"Total rows: {len(fao_df)}")
fao_df.head()

Total rows: 4220


,document_id,title,topic,crop,agro_zone,text,origin,source_url,license
0,fao_001_0001,Climate Change and Food Systems Resilience in ...,climate_adaptation,multiple,sub_humid,CLIMATEC HANGE AND FOOD SYSTEMS RESILIENCEI N ...,fao,https://www.fao.org/4/i2230e/i2230e.pdf,CC-BY-NC-SA-3.0-IGO
1,fao_001_0002,Climate Change and Food Systems Resilience in ...,climate_adaptation,multiple,sub_humid,imply the expression of any opinion whatsoever...,fao,https://www.fao.org/4/i2230e/i2230e.pdf,CC-BY-NC-SA-3.0-IGO
2,fao_001_0003,Climate Change and Food Systems Resilience in ...,climate_adaptation,multiple,sub_humid,. The mention of specific companies or product...,fao,https://www.fao.org/4/i2230e/i2230e.pdf,CC-BY-NC-SA-3.0-IGO
3,fao_001_0004,Climate Change and Food Systems Resilience in ...,climate_adaptation,multiple,sub_humid,. Non-commercial uses will be authorized free ...,fao,https://www.fao.org/4/i2230e/i2230e.pdf,CC-BY-NC-SA-3.0-IGO
4,fao_001_0005,Climate Change and Food Systems Resilience in ...,climate_adaptation,multiple,sub_humid,".org or to the Chief, Publishing Policy and Su...",fao,https://www.fao.org/4/i2230e/i2230e.pdf,CC-BY-NC-SA-3.0-IGO


In [ ]:
boilerplate_keywords = [
    "non-commercial uses", "copyright", "all rights reserved",
    "reproduction", "chief publishing", "chief, publishing",
    "mention of specific companies", "expression of any opinion",
    "food and agriculture organization", "climate change and food systems resiliencei",
    "imply the expression", "authorized free", "preface",
    ".org or to the", "this volume is published", "editorial note",
    "table of contents", "list of figures", "list of tables",
    "acknowledgement", "foreword", "abbreviations"
]

def is_boilerplate(text):
    text_lower = text.lower()
    return any(kw in text_lower for kw in boilerplate_keywords)

before = len(fao_df)
fao_df = fao_df[~fao_df['text'].apply(is_boilerplate)].reset_index(drop=True)

fao_df['text'] = fao_df['text'].str.replace(r'^[\s\.,"]+', '', regex=True).str.strip()

fao_df = fao_df[fao_df['text'].str.len() >= 60].reset_index(drop=True)

print(f"Removed {before - len(fao_df)} boilerplate rows")
print(f"Final clean rows: {len(fao_df)}")

fao_df.head()

Removed 68 boilerplate rows
Final clean rows: 4152


,document_id,title,topic,crop,agro_zone,text,origin,source_url,license
0,fao_001_0007,Climate Change and Food Systems Resilience in ...,climate_adaptation,multiple,sub_humid,"Nevertheless, it was fast enough to feed a gro...",fao,https://www.fao.org/4/i2230e/i2230e.pdf,CC-BY-NC-SA-3.0-IGO
1,fao_001_0008,Climate Change and Food Systems Resilience in ...,climate_adaptation,multiple,sub_humid,This loss of the genetic wealth of our biosphe...,fao,https://www.fao.org/4/i2230e/i2230e.pdf,CC-BY-NC-SA-3.0-IGO
2,fao_001_0009,Climate Change and Food Systems Resilience in ...,climate_adaptation,multiple,sub_humid,"Only lately, just two or three centuries ago, ...",fao,https://www.fao.org/4/i2230e/i2230e.pdf,CC-BY-NC-SA-3.0-IGO
3,fao_001_0010,Climate Change and Food Systems Resilience in ...,climate_adaptation,multiple,sub_humid,The fast rate of changing environmental condit...,fao,https://www.fao.org/4/i2230e/i2230e.pdf,CC-BY-NC-SA-3.0-IGO
4,fao_001_0011,Climate Change and Food Systems Resilience in ...,climate_adaptation,multiple,sub_humid,"In following decades, this acceptance of homog...",fao,https://www.fao.org/4/i2230e/i2230e.pdf,CC-BY-NC-SA-3.0-IGO


In [ ]:
fao_df['text'][3]

'The fast rate of changing environmental conditions, namely temperature and rainwater levels and variability, could no longer be matched by species adaptation.  iii  The industrial revolution also led to mass production, because of the ease with which the new machines could spew out identical copies of tools and furniture and cars, more efficiently and cheaply than handmade versions'

In [ ]:
fao_df['text'][4]

'In following decades, this acceptance of homogeneity spread from factories into fields. Agricultures focus shifted to commercial crop varieties with their improved output, at the loss of those handmade or indigenous varieties adapted to unique environments and resistant to pests and diseases'

In [ ]:
fao_df['text'][2000]

'The establishment of this restaurant aims at initiating a farm to restaurant marketing strategy, where organic producers and consumers can meet. Therefore, the growth of local markets for organic produce can be triggered. Café Baobab was formally opened in July 2007 with the help of GOAN, by Lady Tangi, an Australian whose husband is a Ghanaian'

In [ ]:
fao_df['text'][400]

'68  INTERNATIONAL ASSESSMENT OF AGRICULTURAL KNOwLEDGE, SCIENCE AND TECHNOLOGY FOR DEvELOPMENT: SUMMARY FOR DECISION MAKERS OF THE SUB-SAHARAN AFRICA REPORT Financial support was provided to the IAASTD by the cosponsoring agencies, the governments of Australia, Canada, Finland, France, Ireland, Sweden, Switzerland, US and UK, the European Commission, and CropLife International'

In [ ]:
fao_df['text'][5]

'The result: erosion of agricultural biodiversity which, in turn, has weakened the ability of agro-ecosystems to adapt to climate change and increased weather variability. There is absolutely no doubt that there is need to increase global food production to meet the demand of a growing population'

In [ ]:
from google.colab import drive

drive.mount('/content/drive', force_remount=False)

import os
import pandas as pd

out_dir = "/content/drive/MyDrive/agriculture_rag"
os.makedirs(out_dir, exist_ok=True)

fao_df = fao_df[fao_df['text'].str.len() >= 80].reset_index(drop=True)

fao_df = fao_df[~fao_df['text'].str.contains(r"people no\.", regex=True)].reset_index(drop=True)

print(f"Final rows: {len(fao_df)}")

out_path = f"{out_dir}/fao_documents.csv"
fao_df.to_csv(out_path, index=False)
print(f"✓ Saved {len(fao_df)} rows → {out_path}")

verify_df = pd.read_csv(out_path)

print(f"✓ Verified: {len(verify_df)} rows, {len(verify_df.columns)} columns")
print(verify_df.columns.tolist())

verify_df.head()

Mounted at /content/drive
Final rows: 4097
✓ Saved 4097 rows → /content/drive/MyDrive/agriculture_rag/fao_documents.csv
✓ Verified: 4097 rows, 9 columns
['document_id', 'title', 'topic', 'crop', 'agro_zone', 'text', 'origin', 'source_url', 'license']


,document_id,title,topic,crop,agro_zone,text,origin,source_url,license
0,fao_001_0007,Climate Change and Food Systems Resilience in ...,climate_adaptation,multiple,sub_humid,"Nevertheless, it was fast enough to feed a gro...",fao,https://www.fao.org/4/i2230e/i2230e.pdf,CC-BY-NC-SA-3.0-IGO
1,fao_001_0008,Climate Change and Food Systems Resilience in ...,climate_adaptation,multiple,sub_humid,This loss of the genetic wealth of our biosphe...,fao,https://www.fao.org/4/i2230e/i2230e.pdf,CC-BY-NC-SA-3.0-IGO
2,fao_001_0009,Climate Change and Food Systems Resilience in ...,climate_adaptation,multiple,sub_humid,"Only lately, just two or three centuries ago, ...",fao,https://www.fao.org/4/i2230e/i2230e.pdf,CC-BY-NC-SA-3.0-IGO
3,fao_001_0010,Climate Change and Food Systems Resilience in ...,climate_adaptation,multiple,sub_humid,The fast rate of changing environmental condit...,fao,https://www.fao.org/4/i2230e/i2230e.pdf,CC-BY-NC-SA-3.0-IGO
4,fao_001_0011,Climate Change and Food Systems Resilience in ...,climate_adaptation,multiple,sub_humid,"In following decades, this acceptance of homog...",fao,https://www.fao.org/4/i2230e/i2230e.pdf,CC-BY-NC-SA-3.0-IGO
